internals/daemon/daemon_test.go

```diff
@@ -336,106 +337,140 @@ func (s *daemonSuite) TestFillsWarnings(c *C) {
 func (s *daemonSuite) TestGuestAccess(c *C) {
 	d := s.newDaemon(c)
 
-	get := &http.Request{Method: "GET"}
-	put := &http.Request{Method: "PUT"}
-	pst := &http.Request{Method: "POST"}
-	del := &http.Request{Method: "DELETE"}
-
-	cmd := &Command{d: d}
-	c.Check(cmd.canAccess(get, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(put, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(pst, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(del, nil), Equals, accessUnauthorized)
-
-	cmd = &Command{d: d, AdminOnly: true}
-	c.Check(cmd.canAccess(get, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(put, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(pst, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(del, nil), Equals, accessUnauthorized)
-
-	cmd = &Command{d: d, UserOK: true}
-	c.Check(cmd.canAccess(get, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(put, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(pst, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(del, nil), Equals, accessUnauthorized)
-
-	cmd = &Command{d: d, GuestOK: true}
-	c.Check(cmd.canAccess(get, nil), Equals, accessOK)
-	c.Check(cmd.canAccess(put, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(pst, nil), Equals, accessUnauthorized)
-	c.Check(cmd.canAccess(del, nil), Equals, accessUnauthorized)
-}
+	responseFunc := func(c *Command, r *http.Request, s *UserState) Response {
+		return SyncResponse(true)
+	}
 
-func (s *daemonSuite) TestUserAccess(c *C) {
-	d := s.newDaemon(c)
+	doTestReqFunc := func(c *C, cmd *Command, mth string) *httptest.ResponseRecorder {
+		req := &http.Request{Method: mth, RemoteAddr: ""}
+		rec := httptest.NewRecorder()
+		cmd.ServeHTTP(rec, req)
+		return rec
+	}
+
+	for _, t := range []struct {
+		getStatus, putStatus, postStatus int
+		readAccess, writeAccess          AccessChecker
+	}{
```

Comment:
I think this would be clearer split into two parts: the table of tests (in a variable `tests` or whatever), then a simple `for _, t := range tests { ... }` loop. I think it'd also be clearer with named fields in the definition, because it's not clear which is get/put/post (I would have expected post first, for example). I realize it would make it a bit longer than one per line. Something like:

```go
tests := []struct{
  get, post, put int // expected status for each method
  read, write AccessChecker
}{{
  get:   http.StatusOK,
  post:  http.StatusOK,
  put:   http.StatusOK,
  read:  OpenAccess{},
  write: OpenAccess{},
}, {
  ...
}}
```

You could even name the test struct type (and move it to the top level) so you can reuse it the other test(s).

Reply:
Thanks, that cleaned it up nicely. I left the order in the same order as the methods are defined in `daemon.Command` (GET, PUT, POST), but now with the named fields, it indeed becomes clearer which is which. If you feel strongly about it, I can swap the order of post and put here.

Once that was done, I realized that the "mechanism" is basically the same for all 3 test cases (`TestGuestAccess`, `TestUserAccess` and `TestSuperAccess`), so I added a helper function `testAccessChecker()` that contains the logic, and the test cases then just contain data (table of test case data + the `remoteAddr` string).